packages/web-extension/src/walletManager/SignerManager/SignerManager.ts

```diff
@@ -0,0 +1,155 @@
+    const transaction = Serialization.Transaction.fromCbor(tx);
+    return this.#signRequest(
+      this.transactionWitnessRequest$,
+      {
+        requestContext,
+        signContext,
+        transaction,
+        walletType: requestContext.wallet.type
+      },
+      (keyAgent) =>
+        keyAgent.signTransaction(
+          {
+            body: transaction.body().toCore(),
+            hash: transaction.getId()
+          },
+          signContext,
+          options
+        )
+    );
+  }
+
+  async signData(props: SignDataProps, requestContext: RequestContext<WalletMetadata>): Promise<SignBlobResult> {
+    return this.#signRequest(
+      this.signDataRequest$,
+      {
+        ...props,
+        requestContext,
+        walletType: requestContext.wallet.type
+      },
+      (keyAgent) => keyAgent.signBlob(props.derivationPath, props.blob)
+    );
+  }
+
+  #signRequest<R, Req extends RequestBase<WalletMetadata> & SignRequest<R>>(
+    emitter$: Subject<Req>,
+    request: Omit<Req, 'reject' | 'sign'>,
+    sign: (keyAgent: KeyAgent) => Promise<R>
+  ) {
+    return new Promise<R>((resolve, reject) => {
+      if (!emitter$.observed) {
+        return reject(new errors.AuthenticationError('Internal error: signDataRequest$ not observed'));
+      }
+      const account = request.requestContext.wallet.accounts.find(
+        ({ accountIndex }) => accountIndex === request.requestContext.accountIndex
+      );
+      if (!account) {
+        return reject(new errors.ProofGenerationError(`Account not found: ${request.requestContext.accountIndex}`));
+      }
+      const bubbleResolveReject = async (action: () => Promise<R>): Promise<R> => {
+        try {
+          const result = action();
+          resolve(result);
+          return result;
+        } catch (error) {
+          reject(error);
+          throw error;
+        }
+      };
+      const commonRequestProps = {
+        ...request,
+        reject: async (reason: string) => reject(new errors.AuthenticationError(reason))
+      };
+      emitter$.next(
+        request.walletType === WalletType.InMemory
+          ? ({
+              ...commonRequestProps,
+              sign: async (passphrase: Uint8Array) =>
+                bubbleResolveReject(() => {
+                  const wallet = request.requestContext.wallet as InMemoryWallet<WalletMetadata>;
+                  return sign(
+                    this.#keyAgentFactory.InMemory({
+                      accountIndex: account.accountIndex,
+                      chainId: request.requestContext.chainId,
+                      encryptedRootPrivateKeyBytes: [
+                        ...Buffer.from(wallet.encryptedSecrets.rootPrivateKeyBytes, 'hex')
+                      ],
+                      extendedAccountPublicKey: wallet.extendedAccountPublicKey,
+                      // TODO: this might be in memory for longer than needed
+                      getPassphrase: async () => passphrase
```

Comment:
This one makes me a bit uneasy 😅 , is there a way to profile this to understand how long would this stay as plain data on memory?

Reply:
a2b88eb91fc1053e6a1dec7c1d2f44ae8046b5a4 should mitigate the concern